Generate Windows-compatible OpenSSH config in `tsh config`

timothyb89 opened the pull request:
This tweaks `tsh config` to generate OpenSSH config blocks compatible with Windows. It works around several issues:
 * Hosts must be translated from a full hostname (e.g. `node.foo.example.com`) to a Teleport node name (`node`) to work consistently. On Unix clients we can use a bash subshell snippet to extract the cluster domain but this isn't possible on Windows. Instead, this adds a hidden tsh subcommand (`tsh config-proxy`) to act as a `ProxyCommand` that manipulates the strings as necessary.
 * Windows does not have an ssh-agent enabled by default. This configures `IdentityFile` and `CertificateFile` so no ssh-agent is needed. This should also improve the experience for users without a compatible ssh-agent (e.g. GNOME).
 * Windows requires a full executable path in `ProxyCommand` directives.

This will improve support for VSCode on Windows (#7444).

timothyb89 committed:
Generate Windows-compatible OpenSSH config in `tsh config`

This tweaks `tsh config` to generate OpenSSH config blocks compatible
with Windows. It works around several issues:
 * Hosts must be translated from a full hostname (e.g.
   `node.foo.example.com`) to a Teleport node name (`node`). On Unix
   clients we can use a bash subshell snippet to extract the cluster
   domain but this isn't possible on Windows. Instead, this adds a
   hidden tsh subcommand (`tsh config-proxy`) to act as a
   `ProxyCommand` that manipulates the strings as necessary.
 * Windows does not have an ssh-agent enabled by default. This
   configures `IdentityFile` and `CertificateFile` so no ssh-agent
   is needed. This should also improve the experience for users
   without a compatible ssh-agent (e.g. GNOME).
 * Windows requires a full executable path in `ProxyCommand`
   directives.

timothyb89 committed:
Remove unnecessary conversion

timothyb89 commented:
I'm not particularly tied to the particular naming of this new tsh subcommand but ugly as it may be, I think some ProxyCommand wrapper is important for a good experience on Windows. I expect this string manipulation to still be necessary in the future but would like to see this make use of the new proxy work in #7524 once that's ready.

awly commented on tool/tsh/config.go:
```diff
@@ -19,44 +19,64 @@ package main
-    {{- if .leaf }}
-    ProxyCommand ssh -p {{ .proxyPort }} {{ .proxyHost }} -s proxy:$(echo %h | cut -d '.' -f 1):%p@{{ .clusterName }}	
+    {{- if .Leaf }}
+    ProxyCommand {{ .TSHPath }} config-proxy --proxy={{ .ProxyHost }}:{{ .ProxyPort }} %h:%p "{{ .ClusterName }}"
```
Can `ProxyCommand` be `tsh ssh %h`?
We could teach `tsh ssh` to do any string parsing to get the node name and would avoid a new subcommand.

timothyb89 replied:
`tsh ssh` doesn't appear to support the subsystem (`-s`) flag, I'm not sure how much effort that would take to implement. `tsh proxy ssh` from https://github.com/gravitational/teleport/pull/7524 would do what we need, other than the string parsing.

My only concern with integrating string parsing into another command is that it will change the hostname semantics. It's not safe to always trim off everything after the first `.`, and it would be challenging to determine whether or not it should be trimmed automatically. Maybe some `--trim-host-suffix` flag could be added to `tsh proxy ssh`?

@smallinsky any thoughts on this?

awly replied:
We don't have to respect the same flags as `ssh` does.
IIUC, the only difference between the current `tsh ssh` behavior and what `ProxyCommand` needs is not messing with the terminal (putting it into raw mode), and simply piping stdin/stdout.

As for hostname interpretation, we can also probe the auth API to see if a node exists.
Run the `GetNode` call with the hostname as is and with `.ClusterName` suffix trimmed and see if either exists.

timothyb89 replied:
> IIUC, the only difference between the current `tsh ssh` behavior and what `ProxyCommand` needs is not messing with the terminal (putting it into raw mode), and simply piping stdin/stdout.

As I understand it, `ProxyCommand` tunnels the connection to the sshd itself, after which the local ssh client begins to communicate with the remote sshd using the regular ssh protocol. On the other hand, `tsh ssh` delivers us an already-connected channel to some remote program (e.g. bash) on the target node.

@russjones figures it might be possible to tweak `tsh ssh` to instead (optionally) open a proxied connection to just the  target sshd, but that's exactly the functionality we'll get from #7524. We could spawn a remote `sshd -i` or something, but that's getting ugly fast (and requires permissions and configuration).

The hope here is to implement this as a dumb layer on top of `ssh` for now, but then swap the implementation out to use #7524's `LocalProxy.SSHProxy()` under the hood instead of a subprocess once it's ready.

awly replied:
`tsh proxy ssh` seems like a better overall solution, but it won't be ready for a while.
Let's use this PR as a stopgap and leave a `TODO` to switch to `tsh proxy ssh` later on.

awly commented on tool/tsh/config.go:
```diff
@@ -121,3 +160,39 @@ func onConfig(cf *CLIConf) error {
+		"-s",
+	}
+
+	if cf.ConfigProxyLeaf != "" {
+		args = append(args, fmt.Sprintf("proxy:%s:%s@%s", targetHost, targetPort, cf.ConfigProxyLeaf))
+	} else {
+		args = append(args, fmt.Sprintf("proxy:%s:%s", targetHost, targetPort))
+	}
```
I think we can always include the cluster name, even when it's root
You should be able to remove some conditional branches this way

timothyb89 replied:
That's much nicer, done!

xacrimon commented on tool/tsh/config.go:
```diff
@@ -19,44 +19,64 @@ package main
+		return "C:\\Windows\\System32\\OpenSSH\\ssh.exe"
+	}
+
+	return "ssh"
```
This seems to be relying on PATH or shell evaluation somehow. That said I think POSIX mandates `/usr/bin/ssh` so that should be a sane default.

timothyb89 replied:
That makes sense, fixed.

timothyb89 committed:
Use /usr/bin/ssh explicitly in `tsh config` template for Unix

timothyb89 committed:
Remove special case for leaf clusters; always require a SiteName

timothyb89 commented on tool/tsh/tsh.go:
```diff
@@ -242,6 +242,12 @@ type CLIConf struct {
+	ConfigProxyTarget string
+
+	// ConfigProxyLeaf is the name of the leaf cluster, if any.
+	ConfigProxyLeaf string
```
There were potential semantic differences before but now that a cluster name is required it just uses `cf.SiteName`

awly commented on tool/tsh/config.go:
```diff
@@ -105,13 +121,29 @@ func onConfig(cf *CLIConf) error {
+			ClusterName:         leafCluster.GetName(),
+			KnownHostsPath:      knownHostsPath,
+			IdentityFilePath:    identityFilePath,
+			CertificateFilePath: keypaths.SSHCertPath(keysDir, proxyHost, tc.Config.Username, rootClusterName),
```
should this use `leafCluster.GetName()` instead of `rootClusterName`?

timothyb89 replied:
I believe the root cluster's is the only cert we can guarantee will exist, tsh doesn't cache certs for leaf clusters unless specifically asked. The leaf clusters trust should all trust the root cert anyway, so it should work fine.

awly commented on tool/tsh/config.go:
```diff
@@ -121,3 +153,35 @@ func onConfig(cf *CLIConf) error {
+		proxyPort,
+		proxyHost,
+		"-s",
+	}
+
+	args = append(args, fmt.Sprintf("proxy:%s:%s@%s", targetHost, targetPort, cf.SiteName))
```
```suggestion
		fmt.Sprintf("proxy:%s:%s@%s", targetHost, targetPort, cf.SiteName),
	}
```

awly commented on tool/tsh/tsh.go:
```diff
@@ -501,6 +504,13 @@ func Run(args []string, opts ...cliOption) error {
+	// `tsh config`. As it's not intended to run by users directly and may
+	// not have a stable CLI interface, hide it.
+	configProxy := app.Command("config-proxy", "ProxyCommand wrapper for SSH config as generated by `config`").Hidden()
+	configProxy.Arg("target", "Target node name").Required().StringVar(&cf.ConfigProxyTarget)
```
```suggestion
	configProxy.Arg("target", "Target node host:port").Required().StringVar(&cf.ConfigProxyTarget)
```

awly commented on tool/tsh/tsh.go:
```diff
@@ -501,6 +504,13 @@ func Run(args []string, opts ...cliOption) error {
+	// not have a stable CLI interface, hide it.
+	configProxy := app.Command("config-proxy", "ProxyCommand wrapper for SSH config as generated by `config`").Hidden()
+	configProxy.Arg("target", "Target node name").Required().StringVar(&cf.ConfigProxyTarget)
+	configProxy.Arg("cluster-name", "Cluster bane").Required().StringVar(&cf.SiteName)
```
```suggestion
	configProxy.Arg("cluster-name", "Target cluster name").Required().StringVar(&cf.SiteName)
```

timothyb89 replied:
Well, that's just an embarrassing typo on my part. Thanks!

timothyb89 committed:
Apply suggestions from code review

Co-authored-by: Andrew Lytvynov <andrew@goteleport.com>

xacrimon reviewed (approved):
lgtm

timothyb89 commented:
I think based on https://github.com/gravitational/teleport/pull/7984 and https://github.com/gravitational/teleport/pull/8088 we'll need to pass through the remote user as well. I'll work on that...

timothyb89 committed:
Pass through remote login name

This should improve compatibility with OIDC and other users with
federated Teleport usernames. The teleport proxy should always accept
a remote username for which the user's certificate is valid.

timothyb89 commented:
Just made a small change to pass the remote login name through to the `ssh` subprocess, this should absorb the changes in  #7984 and in #8088.

russjones reviewed (approved):
One comment, other than that looks good to me.

timothyb89 committed:
Use `exec.LookPath` to resolve the ssh path

This prefers whichever `ssh` exists on the PATH for all OSes. After some
testing, Git for Windows SSH works just as well as Microsoft's, so we don't
need to overspecify things.

Also, quotes the tsh.exe path in generated config. Git for Windows' ssh
didn't autoescape the Windows paths.

timothyb89 commented on tool/tsh/config.go:
```diff
@@ -19,44 +19,59 @@ package main
+}
+
+// getSSHPath returns a sane default `ssh` path for the current platform.
+func getDefaultSSHPath() string {
```
I've switched over to using `LookPath` on all platforms, after some testing the main other SSH binary we're likely to see on Windows (from Git for Windows) works perfectly fine. This should make Git for Windows a viable client on older Windows versions since #7790 only supports Win10.

timothyb89 committed:
Merge branch 'master' into timothyb89/tsh-config-windows
